scripts/pylib/twister/twisterlib.py

```diff
@@ -649,25 +649,22 @@ def monitor_serial(self, ser, halt_fileno, harness):
 
         log_out_fp.close()
 
-    def get_available_device(self, instance):
-        device = instance.platform.name
-        for d in self.suite.duts:
-            if d.platform == device and d.available and (d.serial or d.serial_pty):
-                d.available = 0
-                d.counter += 1
-                return d
-
-        return None
-
     def device_is_available(self, instance):
         device = instance.platform.name
         fixture = instance.testcase.harness_config.get("fixture")
         for d in self.suite.duts:
             if fixture and fixture not in d.fixtures:
                 continue
-            if d.platform == device and d.available and (d.serial or d.serial_pty):
+            if d.platform != device or not (d.serial or d.serial_pty):
+                continue
+            d.lock.acquire()
```

Comment:
both d.available and d.counter are already protected by a lock, not sure why it was failing..

```
    @available.setter
    def available(self, value):
        with self._available.get_lock():
            self._available.value = value

    @property
    def counter(self):
        with self._counter.get_lock():
            return self._counter.value
```

Reply:
Ah, didn't notice that.  We should be using that lock object.  But it's not enough to protect the one access, you have to protect the sequence where you check the value, make a decision, and update the value.  Example:

* Thread A checks "available", sees that it is 1, sets it to 0, and begins executing a test.  This is what we expect to happen.  All other threads are still polling on available to become 1 and don't do anything.

* Thread A finishes the test and sets avaialble to 1

* SWITCH!

* Thread B sees that available is 1 and ...

* SWITCH!

* Thread C sees that available is 1, sets it to zero, and begins executing a test.

* SWITCH!

* Thread B sets available to zero, and begins executing a test.

Now we have to threads running the test.  The decision about the old value of available, the action to be taken, and the new value, all need to be atomic from the perspective of other threads.